src/NServiceBus.AcceptanceTests/Core/Pipeline/When_modifying_headers_during_forwarding.cs

```diff
@@ -0,0 +1,92 @@
+﻿namespace NServiceBus.AcceptanceTests.Core.Pipeline
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using AcceptanceTesting;
+    using EndpointTemplates;
+    using NServiceBus.Pipeline;
+    using NUnit.Framework;
+
+    public class When_modifying_headers_during_forwarding : NServiceBusAcceptanceTest
+    {
+        [Test]
+        public async Task Should_not_leak_changes_into_main_pipeline()
+        {
+            var context = await Scenario.Define<Context>()
+                .WithEndpoint<ForwardingEndpoint>(e => e
+                    .When(s => s.SendLocal(new SomeMessage())))
+                .Done(c => c.HeadersInHandler != null)
+                .Run();
+
+            Assert.IsFalse(context.HeadersInHandler.ContainsKey("ForwardingHeader"));
+            Assert.IsFalse(context.HeadersInPipeline.ContainsKey("ForwardingHeader"));
+        }
+
+        class Context : ScenarioContext
+        {
+            public IReadOnlyDictionary<string, string> HeadersInHandler { get; set; }
+            public IReadOnlyDictionary<string, string> HeadersInPipeline { get; set; }
+        }
+
+        class ForwardingEndpoint : EndpointConfigurationBuilder
+        {
+            public ForwardingEndpoint()
+            {
+                EndpointSetup<DefaultServer>(e =>
+                {
+                    e.ForwardReceivedMessagesTo("forwardingQueue");
+                    e.Pipeline.Register(typeof(ForwardingBehavior), "Adds headers to a forwarded message");
+                    e.Pipeline.Register(typeof(MainPipelineBehavior), "Adds headers to a forwarded message");
+                });
+            }
+
+            class SomeMessageHandler : IHandleMessages<SomeMessage>
+            {
+                Context testContext;
+
+                public SomeMessageHandler(Context testContext)
+                {
+                    this.testContext = testContext;
+                }
+
+                public Task Handle(SomeMessage message, IMessageHandlerContext context)
+                {
+                    testContext.HeadersInHandler = context.MessageHeaders.ToDictionary(x => x.Key, x => x.Value);
```

Comment:
Not going to propose this change, but I'd rather instead of copying the headers and asserting in the test, would have a boolean on the `testContext` to indicate wherever the header was present or not.

Reply:
that would work as well. Not sure I see a clear benefit from one over the other approach tbh.

Reply:
I can see pros of both approaches ;-) The current one makes it more visible what the test is about just by reading the test method so I am 👍 for leaving it as is.